### PR TITLE
Added support for Apple Pay

### PR DESCRIPTION
**Pull request by clintonb**
LEARNER-1668

- [x] Finish unit tests
- [x] Decide on whether to add e2e test (with mocked Apple Pay endpoints)
- [x] Site configuration flag to disable
- [x] Finish JS tests
- [x] Increase overall test coverage

### CODE REVIEW

**Comment by clintonb** [comment]
@edx/learner please review

File: ecommerce/extensions/payment/processors/cybersource.py [hunk 1]
```diff
@@ -348,3 +357,99 @@ def issue_credit(self, order, reference_number, amount, currency):
                 'Failed to issue CyberSource credit for order [{order_number}]. '
                 'Complete response has been recorded in entry [{response_id}]'.format(
                     order_number=order.number, response_id=ppr.id))
+
+    def request_apple_pay_authorization(self, basket, billing_address, payment_token):
```
Comment (mjfrey) [inline ecommerce/extensions/payment/processors/cybersource.py:362]: Why attach Apple Pay to Cybersource payment processor?  What if we move or add a new processor?

Comment (clintonb) [reply ecommerce/extensions/payment/processors/cybersource.py:362]: Making a generic Apple Pay setup for the first pass is overkill. I thought about it, but the implementations are somewhat processor-specific. Stripe, for example does much of the work for merchants when starting the session. What's written is specific to CyberSource for the simple fact that making it generic would be a bit of a hack.

**Comment by clintonb** [comment]
Related: https://github.com/edx/edx-themes/pull/528

File: ecommerce/extensions/payment/processors/cybersource.py [hunk 2]
```diff
@@ -348,3 +357,99 @@ def issue_credit(self, order, reference_number, amount, currency):
...
+                'postalCode': billing_address.postcode,
+                'country': billing_address.country.iso_3166_1_a2,
+                'email': basket.owner.email,
+            }
+            purchase_totals = {
+                'currency': basket.currency,
+                'grandTotalAmount': str(basket.total_incl_tax),
+            }
+            encrypted_payment = {
+                'descriptor': 'RklEPUNPTU1PTi5BUFBMRS5JTkFQUC5QQVlNRU5U',
```
Comment (rlucioni) [inline ecommerce/extensions/payment/processors/cybersource.py:399]: What's this? It looks like hardcoded configuration.

File: ecommerce/extensions/payment/processors/cybersource.py [hunk 3]
```diff
@@ -348,3 +357,99 @@ def issue_credit(self, order, reference_number, amount, currency):
...
+            response = client.service.runTransaction(
+                merchantID=self.merchant_id,
+                merchantReferenceCode=basket.order_number,
+                billTo=bill_to,
+                purchaseTotals=purchase_totals,
+                encryptedPayment=encrypted_payment,
+                card=card,
+                ccAuthService=auth_service,
+                ccCaptureService=capture_service,
+                paymentSolution='001',
```
Comment (rlucioni) [inline ecommerce/extensions/payment/processors/cybersource.py:431]: Does this belong in settings?

File: ecommerce/settings/_oscar.py [hunk 4]
```diff
@@ -131,6 +132,13 @@
 }
 
 PAYMENT_PROCESSOR_SWITCH_PREFIX = 'payment_processor_active_'
+
+# TODO Move to CyberSource settings
```
Comment (rlucioni) [inline ecommerce/settings/_oscar.py]: Looks like you took care of this TODO? I don't see these settings being used anywhere.

File: ecommerce/static/js/payment_processors/cybersource.js [hunk 5]
```diff
@@ -123,6 +127,98 @@ require([
...
+                    },
+                    data: JSON.stringify({url: evt.validationURL}),
+                    contentType: 'application/json',
+                    success: function(data) {
+                        console.log('Merchant validation succeeded.');
+                        console.log(data);
+                        session.completeMerchantValidation(data);
+                    },
+                    error: function(jqXHR, textStatus, errorThrown) {
+                        // TODO Display an error message to the learner
```
Comment (rlucioni) [inline ecommerce/static/js/payment_processors/cybersource.js]: Should these console logs be removed?

File: ecommerce/urls.py [hunk 6]
```diff
@@ -40,7 +41,13 @@ def handler403(_):
 # NOTE 2: These same patterns are used for rest_framework's browseable API authentication links.
 AUTH_URLS = [url(r'^logout/$', LogoutView.as_view(), name='logout'), ] + auth_urlpatterns
 
-urlpatterns = AUTH_URLS + [
+WELL_KNOWN_URLS = [
+    # TODO Add view for /.well-known/apple-developer-merchantid-domain-association
```
Comment (rlucioni) [inline ecommerce/urls.py]: Looks like you took care of this, too.

**Review by rlucioni — COMMENTED** [review]
@clintonb would it make sense to add a section to the docs for this service that gives a high level overview of how Apple Pay works and how to enable it?

File: ecommerce/extensions/payment/processors/cybersource.py (hunk 2 above)
Comment (clintonb) [reply ecommerce/extensions/payment/processors/cybersource.py:399]: That is intentional. It's required by CyberSource.

File: ecommerce/extensions/payment/processors/cybersource.py (hunk 3 above)
Comment (clintonb) [reply ecommerce/extensions/payment/processors/cybersource.py:431]: That is intentional. It's required by CyberSource.

File: ecommerce/extensions/payment/processors/cybersource.py (hunk 2 above)
Comment (rlucioni) [reply ecommerce/extensions/payment/processors/cybersource.py:399]: @clintonb a comment explaining what this is would be helpful to myself and other maintainers who aren't as familiar with this.

Comment (clintonb) [reply ecommerce/extensions/payment/processors/cybersource.py:399]: Hmm...how about a link to the CyberSource docs in the method's docstring? Candidly, if folks don't know what they are doing—haven't read the docs—they shouldn't be messing around in payment code.

Comment (rlucioni) [reply ecommerce/extensions/payment/processors/cybersource.py:399]: @clintonb sure, that link would be good.

Comment (clintonb) [reply ecommerce/extensions/payment/processors/cybersource.py:399]: Docs and docstring updated.

File: ecommerce/static/js/payment_processors/cybersource.js (hunk 5 above)
Comment (clintonb) [reply ecommerce/static/js/payment_processors/cybersource.js]: I find them useful for development. They don't affect the end user, so I see no great need to remove them.

**Comment by clintonb** [comment]
@rlucioni @mjfrey please take another look when you can. I have higher priority issues, so it will take some time for me to fix the coverage issues and the last test, however, the server side components are ready to go.

File: .gitignore [hunk 7]
```diff
@@ -92,3 +92,7 @@ docs/_build
 
 # Visual Studio Code
 .vscode
+
+# Certificates
+*.cer
+*.pem
```
Comment (rlucioni) [inline .gitignore:98]: @clintonb are there Ansible changes corresponding to this PR that are necessary to manage this certificate?

**Review by rlucioni — APPROVED** [review]

File: .gitignore (hunk 7 above)
Comment (clintonb) [reply .gitignore:98]: Yes.

**Comment by clintonb** [comment]
Related PR for certificates: https://github.com/edx/configuration/pull/3990

**Pull request merged**